rules/alerting_test.go

```diff
@@ -434,6 +434,65 @@ func TestAlertingRuleEmptyLabelFromTemplate(t *testing.T) {
 	require.Equal(t, result, filteredRes)
 }
 
+func TestAlertingRuleQueryInTemplate(t *testing.T) {
+	suite, err := promql.NewTest(t, `
+		load 1m
+			http_requests{job="app-server", instance="0"}	70 85 70 70
+	`)
+	require.NoError(t, err)
+	defer suite.Close()
+
+	require.NoError(t, suite.Run())
+
+	expr, err := parser.ParseExpr(`sum(http_requests) < 100`)
+	require.NoError(t, err)
+
+	ruleWithQueryInTemplate := NewAlertingRule(
+		"ruleWithQueryInTemplate",
+		expr,
+		time.Minute,
+		labels.FromStrings("label", "value"),
+		labels.FromStrings("templated_label", `{{- with "sort(sum(http_requests) by (instance))" | query -}}
+{{- range $i,$v := . -}}
+instance: {{ $v.Labels.instance }}, value: {{ printf "%.0f" $v.Value }};
+{{- end -}}
+{{- end -}}
+`),
+		nil,
+		"",
+		true, log.NewNopLogger(),
+	)
+	evalTime := time.Unix(0, 0)
+
+	startQueryCh := make(chan struct{})
+	getDoneCh := make(chan struct{})
+	slowQueryFunc := func(ctx context.Context, q string, ts time.Time) (promql.Vector, error) {
+		if q == "sort(sum(http_requests) by (instance))" {
+			// This is minimum produce for issue 10703, expand template with query.
+			t.Logf("q:%s", q)
+			close(startQueryCh)
+			time.Sleep(time.Millisecond * 10)
```

Comment:
Why is the sleep necessary here, is it in order to let the goroutine execute (from line 480)?

Reply:
This is reproduce for expand template with a slow query, it will hold the lock, block the get operation like `.GetEvaluationDuration()` `.LastError()`.

Reply:
I Add a assert to this unit test, it will fail in current master branch, and ok in this pr, ensure solve this issue. you can take a look. :).

Reply:
@hanjm could this be solved by instead waiting on `getDoneCh` here, to ensure that the goroutine finishes before the "slow" evaluation continues?

At least, when I try to make this change, the test still succeeds. Also, when undoing your fix, the test still hangs (as I assume it should).

Reply:
@aknuds1 I think it is impossible. could you show the code?

Reply:
@hanjm this is my diff:

```
diff --git a/rules/alerting_test.go b/rules/alerting_test.go
index 941f455a4..4417b442b 100644
--- a/rules/alerting_test.go
+++ b/rules/alerting_test.go
@@ -466,31 +466,26 @@ instance: {{ $v.Labels.instance }}, value: {{ printf "%.0f" $v.Value }};
 
 	startQueryCh := make(chan struct{})
 	getDoneCh := make(chan struct{})
-	slowQueryDuration := time.Millisecond * 10
 	slowQueryFunc := func(ctx context.Context, q string, ts time.Time) (promql.Vector, error) {
 		if q == "sort(sum(http_requests) by (instance))" {
 			// This is a minimum reproduction of issue 10703, expand template with slow query.
 			close(startQueryCh)
-			time.Sleep(slowQueryDuration)
+			<-getDoneCh
 		}
 		return EngineQueryFunc(suite.QueryEngine(), suite.Storage())(ctx, q, ts)
 	}
 	go func() {
 		<-startQueryCh
-		start := time.Now()
 		_ = ruleWithQueryInTemplate.Health()
 		_ = ruleWithQueryInTemplate.LastError()
 		_ = ruleWithQueryInTemplate.GetEvaluationDuration()
 		_ = ruleWithQueryInTemplate.GetEvaluationTimestamp()
 		close(getDoneCh)
-		// Assert not block when template expanding.
-		require.True(t, time.Since(start) < slowQueryDuration)
 	}()
 	_, err = ruleWithQueryInTemplate.Eval(
 		suite.Context(), evalTime, slowQueryFunc, nil, 0,
 	)
 	require.NoError(t, err)
-	<-getDoneCh
 }
 
 func TestAlertingRuleDuplicate(t *testing.T) {
```

The way I understand it, it should accomplish the same thing, by making `slowQueryFunc` wait while the goroutine executes.

Reply:
Good idea, i will update it later. Thank you.

Reply:
@aknuds1 I made some improvement in latest commit, make the unit test fail in current mater branch instead of hangs.

```
	slowQueryFunc := func(ctx context.Context, q string, ts time.Time) (promql.Vector, error) {
		if q == "sort(sum(http_requests) by (instance))" {
			// This is a minimum reproduction of issue 10703, expand template with query.
			close(startQueryCh)
			select {
			case <-getDoneCh:
			case <-time.After(time.Millisecond * 10):
				// Assert the read goroutine no blocking when template expanding.
				require.Fail(t,"unexpected blocking when template expanding.")
			}
		}
		return EngineQueryFunc(suite.QueryEngine(), suite.Storage())(ctx, q, ts)
	}
```